sound/soc/sof/ipc4-loader.c

```diff
@@ -423,6 +423,62 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+/**
+ * sof_ipc4_update_basecfg_from_manifest - Update the base config from manifest
+ * @sdev: SOF device
+ * @fw_module: pointer struct sof_ipc4_fw_module to parse
+ * @basecfg: Pointer to the base_config to update
+ *
+ * Note: currently only used to update the CPC value if it is not set from
+ * topology already.
+ */
+void sof_ipc4_update_basecfg_from_manifest(struct snd_sof_dev *sdev,
+					   struct sof_ipc4_fw_module *fw_module,
+					   struct sof_ipc4_base_module_cfg *basecfg)
+{
+	const struct sof_man4_module_config *fm_config;
+	u32 cpc_pick = 0;
+	u32 max_cpc = 0;
+	int i;
+
+	/* The CPC is set from topology, ignoring manifest */
+	if (basecfg->cpc)
+		return;
+
+	if (!fw_module || !fw_module->fm_config)
+		return;
+
+	fm_config = fw_module->fm_config;
+	for (i = 0; i < fw_module->man4_module_entry.cfg_count; i++) {
+		if (basecfg->obs == fm_config[i].obs &&
+		    basecfg->ibs == fm_config[i].ibs &&
+		    cpc_pick < fm_config[i].cpc)
+			cpc_pick = fm_config[i].cpc;
+
+		if (max_cpc < fm_config[i].cpc)
+			max_cpc = fm_config[i].cpc;
+	}
+
+	if (!cpc_pick) {
+		dev_dbg(sdev->dev,
+			"%s: No direct CPC match based on obs/ibs %u/%u\n",
+			fw_module->man4_module_entry.name, basecfg->obs, basecfg->ibs);
+		cpc_pick = max_cpc;
+	}
+
+	if (cpc_pick) {
+		dev_dbg(sdev->dev, "%s: CPC from manifest: %u\n",
+			fw_module->man4_module_entry.name, cpc_pick);
+	} else {
```

Comment:
can this branch be reached?

Is there a case where max_cpc == 0?

Reply:
Currently in tgl the manifest only have 0 as CPC among all the configurations, so yes, it can be hit.
In the new version I have changed this as well.